Mathlib/Data/Matroid/Closure.lean

```diff
@@ -0,0 +1,207 @@
+    e ∈ M.cl X ↔ ∀ F, M.Flat F → X ⊆ F → e ∈ F := by
+  simp_rw [cl_eq_sInter_of_subset X, mem_sInter, mem_setOf, and_imp]
+
+lemma subset_cl_iff_forall_subset_flat (X : Set α) (hX : X ⊆ M.E := by aesop_mat) :
+    Y ⊆ M.cl X ↔ ∀ F, M.Flat F → X ⊆ F → Y ⊆ F := by
+  simp_rw [cl_eq_sInter_of_subset X, subset_sInter_iff, mem_setOf, and_imp]
+
+lemma subset_cl (M : Matroid α) (X : Set α) (hX : X ⊆ M.E := by aesop_mat) :
+    X ⊆ M.cl X := by
+  rw [cl_eq_sInter_of_subset X, subset_sInter_iff]
+  simp
+
+lemma Flat.cl (hF : M.Flat F) : M.cl F = F :=
+  (sInter_subset_of_mem (by simpa)).antisymm (M.subset_cl F)
+
+@[simp] lemma cl_ground (M : Matroid α) : M.cl M.E = M.E :=
+  (M.cl_subset_ground M.E).antisymm (M.subset_cl M.E)
+
+@[simp] lemma cl_univ (M : Matroid α) : M.cl univ = M.E := by
+  rw [← cl_inter_ground, univ_inter, cl_ground]
+
+lemma cl_subset_cl (M : Matroid α) (h : X ⊆ Y) : M.cl X ⊆ M.cl Y :=
+  subset_sInter (fun _ ⟨hF, hssF⟩ ↦
+    sInter_subset_of_mem ⟨hF, subset_trans (inter_subset_inter_left _ h) hssF⟩)
+
+lemma cl_mono (M : Matroid α) : Monotone M.cl :=
+  fun _ _ ↦ M.cl_subset_cl
+
+@[simp] lemma cl_cl (M : Matroid α) (X : Set α) : M.cl (M.cl X) = M.cl X :=
+  (M.subset_cl _).antisymm' (subset_sInter
+    (fun F hF ↦ (cl_subset_cl _ (sInter_subset_of_mem hF)).trans hF.1.cl.subset))
+
+lemma cl_subset_cl_of_subset_cl (hXY : X ⊆ M.cl Y) : M.cl X ⊆ M.cl Y :=
+    (M.cl_subset_cl hXY).trans_eq (M.cl_cl Y)
+
+lemma cl_subset_cl_iff_subset_cl (hX : X ⊆ M.E := by aesop_mat) :
+    M.cl X ⊆ M.cl Y ↔ X ⊆ M.cl Y :=
+  ⟨(M.subset_cl X).trans, cl_subset_cl_of_subset_cl⟩
+
+lemma subset_cl_of_subset (M : Matroid α) (hXY : X ⊆ Y) (hY : Y ⊆ M.E := by aesop_mat) :
+    X ⊆ M.cl Y :=
+  hXY.trans (M.subset_cl Y)
+
+lemma subset_cl_of_subset' (M : Matroid α) (hXY : X ⊆ Y) (hX : X ⊆ M.E := by aesop_mat) :
+    X ⊆ M.cl Y := by
+  rw [← cl_inter_ground]; exact M.subset_cl_of_subset (subset_inter hXY hX)
+
+lemma exists_of_cl_ssubset (hXY : M.cl X ⊂ M.cl Y) : ∃ e ∈ Y, e ∉ M.cl X := by
+  by_contra! hcon
+  exact hXY.not_subset (M.cl_subset_cl_of_subset_cl hcon)
+
+lemma mem_cl_of_mem (M : Matroid α) (h : e ∈ X) (hX : X ⊆ M.E := by aesop_mat) :
+    e ∈ M.cl X :=
+  (M.subset_cl X) h
+
+lemma mem_cl_of_mem' (M : Matroid α) (heX : e ∈ X) (h : e ∈ M.E := by aesop_mat) :
+    e ∈ M.cl X := by
+  rw [← cl_inter_ground]; exact M.mem_cl_of_mem ⟨heX, h⟩
+
+lemma not_mem_of_mem_diff_cl (he : e ∈ M.E \ M.cl X) : e ∉ X :=
+  fun heX ↦ he.2 <| M.mem_cl_of_mem' heX he.1
+
+@[aesop unsafe 10% (rule_sets := [Matroid])]
+lemma mem_ground_of_mem_cl (he : e ∈ M.cl X) : e ∈ M.E := (M.cl_subset_ground _) he
+
+lemma cl_iUnion_cl_eq_cl_iUnion (M : Matroid α) (Xs : ι → Set α) :
+    M.cl (⋃ i, M.cl (Xs i)) = M.cl (⋃ i, Xs i) := by
+  refine (M.cl_subset_cl_of_subset_cl
+    (iUnion_subset (fun i ↦ M.cl_subset_cl (subset_iUnion _ _)))).antisymm ?_
+  rw [← cl_inter_ground, iUnion_inter]
+  refine' M.cl_subset_cl (iUnion_subset (fun i ↦ (M.subset_cl _).trans _))
```

Comment:
Why not `refine M.cl_subset_cl (iUnion_subset (fun i ↦ (M.subset_cl _).trans ?_))` ?